node/src/components/chain_synchronizer/operations.rs

```
@@ -1294,7 +1335,135 @@ async fn fetch_block_worker(
+) -> Result<FetchedData<BlockSignatures>, FetcherError<BlockSignatures>> {
+    for _ in 0..retries {
+        let maybe_signatures = ctx
+            .effect_builder
+            .fetch::<BlockSignatures>(block_hash, peer)
+            .await;
+        match maybe_signatures {
+            Ok(result) => return Ok(result),
+            Err(FetcherError::TimedOut { .. }) => continue,
+            Err(_) => return maybe_signatures,
+        }
+    }
+
+    Err(FetcherError::TimedOut {
+        id: block_hash,
+        peer,
+    })
+}
+
+async fn fetch_finality_signatures_by_block_hash(
+    block_hash: BlockHash,
+    validator_weights: &BTreeMap<PublicKey, U512>,
+    ctx: &ChainSyncContext<'_>,
+) -> Option<BlockSignatures> {
+    let mut peer_list = ctx
+        .effect_builder
+        .get_fully_connected_non_joiner_peers()
+        .await;
+    ctx.filter_bad_peers(&mut peer_list);
+
+    let mut sig_collector = BlockSignaturesCollector::new();
+
+    for peer in peer_list {
+        let fetched_signatures = fetch_finality_signatures_with_retry(
+            block_hash,
+            peer,
+            FINALITY_SIGNATURE_FETCH_RETRY_COUNT,
+            ctx,
+        )
+        .await;
+        let maybe_signatures = match fetched_signatures {
+            Ok(FetchedData::FromStorage { item, .. }) => {
+                trace!(
+                    ?block_hash,
+                    ?peer,
+                    "did not get FinalitySignatures from peer, got from storage instead",
+                );
+                Some(*item)
+            }
+            Ok(FetchedData::FromPeer { item, .. }) => {
+                trace!(?block_hash, ?peer, "fetched FinalitySignatures");
+                Some(*item)
+            }
+            Err(_) => None,
+        };
+
+        sig_collector.add_signatures(maybe_signatures);
+
+        if sig_collector
+            .check_if_sufficient(validator_weights, ctx.config.finality_threshold_fraction())
```

We shouldn't add invalid signature(s) - the check is made in `check_if_sufficient` but we add (`add_signatures`) first.

IMO we shouldn't ignore the error returned from `check_if_sufficient` on invalid signature - we should disconnect from the peer.

---

The change is implemented in https://github.com/casper-network/casper-node/pull/3125/commits/46dea0052b116edce0cf317ffbc64f1390815810

However, I introduced the separate validation step, because `check_if_sufficient` actually validates only as a side effect.